src/core/Akka.Persistence/Eventsourced.Recovery.cs

```diff
@@ -124,89 +154,105 @@ private EventsourcedState RecoveryStarted(long maxReplays)
         private EventsourcedState Recovering(Receive recoveryBehavior, TimeSpan timeout)
         {
             // protect against event replay stalling forever because of journal overloaded and such
-            var timeoutCancelable = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(timeout, timeout, Self,
-                new RecoveryTick(false), Self);
+            var timeoutCancelable = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(timeout, timeout, Self, new RecoveryTick(false), Self);
             var eventSeenInInterval = false;
 
             return new EventsourcedState("replay started", true, (receive, message) =>
             {
-                if (message is ReplayedMessage)
+                try
                 {
-                    var m = (ReplayedMessage)message;
-                    try
+                    if (message is ReplayedMessage)
                     {
-                        eventSeenInInterval = true;
-                        UpdateLastSequenceNr(m.Persistent);
-                        base.AroundReceive(recoveryBehavior, m.Persistent);
+                        var m = (ReplayedMessage)message;
+                        try
+                        {
+                            eventSeenInInterval = true;
+                            UpdateLastSequenceNr(m.Persistent);
+                            base.AroundReceive(recoveryBehavior, m.Persistent);
+                        }
+                        catch (Exception cause)
+                        {
+                            timeoutCancelable.Cancel();
+                            try
+                            {
+                                OnRecoveryFailure(cause, m.Persistent.Payload);
+                            }
+                            finally
+                            {
+                                Context.Stop(Self);
+                            }
+                            ReturnRecoveryPermit();
+                        }
                     }
-                    catch (Exception cause)
+                    else if (message is RecoverySuccess)
                     {
+                        var m = (RecoverySuccess)message;
                         timeoutCancelable.Cancel();
+                        OnReplaySuccess();
+                        ChangeState(ProcessingCommands());
+                        _sequenceNr = m.HighestSequenceNr;
+                        LastSequenceNr = m.HighestSequenceNr;
+                        _internalStash.UnstashAll();
+
                         try
                         {
-                            OnRecoveryFailure(cause, m.Persistent.Payload);
+                            base.AroundReceive(recoveryBehavior, RecoveryCompleted.Instance);
                         }
                         finally
                         {
-                            Context.Stop(Self);
+                            ReturnRecoveryPermit();
```

Comment:
No because it's refactored inside the `TransitToProcessingState`. But you need this line, otherwise the whole thing comes to a halt after the first batch of successful recoveries.